src/lib/crypto/snarky_tests/snarky_tests.ml

```diff
@@ -0,0 +1,274 @@
+(* this is ugly, but [Core_kernel.String] doesn't have a [trim] function *)
+let trim = String.trim
+
+open Core_kernel
+
+(* instantiate snarky for Vesta *)
+
+module Impl = Snarky_backendless.Snark.Run.Make (struct
+  include Kimchi_backend.Pasta.Vesta_based_plonk
+  module Inner_curve = Kimchi_backend.Pasta.Pasta.Pallas
+end)
+
+(* helpers *)
+
+let compare_with obtained filepath =
+  let (_ : int) = Sys.command "tree" in
+  let filepath = "examples/" ^ filepath in
+  let expected = In_channel.read_all filepath in
+  if String.(trim obtained <> trim expected) then (
+    Format.printf "mismatch for %s detected\n" filepath ;
+    Format.printf "expected:\n%s\n\n" expected ;
+    Format.printf "obtained:\n%s\n" obtained ;
+    failwith "circuit compilation has changed" )
+
+let check_json ~input_typ ~return_typ ~circuit filename () =
+  let cs : Impl.R1CS_constraint_system.t =
+    Impl.constraint_system ~input_typ ~return_typ circuit
+  in
+  let serialized_json =
+    Kimchi_backend.Pasta.Vesta_based_plonk.R1CS_constraint_system.to_json cs
+  in
+  compare_with serialized_json filename
+
+(* monadic API tests *)
+
+(** Both the monadic and imperative API will produce the same circuit hash. *)
+let expected = "5357346d161dcccaa547c7999b8148db"
+
+module MonadicAPI = struct
```

Comment:
Can you split these up? It would be nice to have them as executables, with some setup like
```
(executable
 (name my-very-fun-test)
 ...
)

(rule
 (alias runtest)
 (action
  (progn
    (with-stdout-to
      my-very-fun-test.out.json
      (run ./my-very-fun-test))
    (diff
      my-very-fun-test.expected.json
      my-very-fun-test.out.json))))
```
for each of the tests, where `*.expected.json` are in the tree.

Reply:
what is the advantage over the current way of doing things with alcotest?

Reply:
A massive file with all of the tests mixed together and a new command line interface to learn to run them both seem undesirable. `dune exec src/lib/crypto/snarky_js_tests/foo.exe` seems more similar to what we do everywhere else, and forces us to split the tests into reasonable units.

Reply:
> A massive file with all of the tests mixed together and a new command line interface to learn to run them both seem undesirable.

I mean we can split things without making multiple executable, and the commands will still run with `dune test` there's nothing that changed (you just get a nicer output if you run it with alcotest)

> More importantly, updating the tests if we do need to is dune runtest src/lib/crypto/snarky_tests; dune promote

I can do something similar without the long rules in a dune file, let's do this in another PR